Minor fixes for artifact removal during map load and handling null map objects

MichalZr6 opened the pull request:
These changes should be the fix for:
- #2166

IvanSavenko commented on lib/mapping/MapFormatH3M.cpp:
```diff
@@ -925,8 +925,7 @@ void CMapLoaderH3M::loadArtifactsOfHero(CGHeroInstance * hero)
 
 		hero->artifactsInBackpack.clear();
-		while(!hero->artifactsWorn.empty())
-			hero->removeArtifact(hero->artifactsWorn.begin()->first);
```
This is wrong.

Heroes can have their artifacts configured in two places:
- globally in scenario options
- in hero instance on map

If hero has artifacts configured in both places, game should only use 2nd set, and discard 1st one

MichalZr6 commented:
So you mean this code loads the wrong set?

I reverted that change and debugged. The issue was a spellbook artifact that was in the backpack. Converting it to `SPELLBOOK_PRESET` during load fixes the map/playability without touching `removeArtifact` but it is still unsafe to use because callbacks (gamestate or EditorCallback) don’t have a fully initialized map yet and `removeArtifact` calls `getMapConstPtr()` along its call stack. I will look into this again later today when I have more time.

```diff
@@ -966,6 +960,13 @@ bool CMapLoaderH3M::loadArtifactToSlot(CGHeroInstance * hero, int slot)
                return false;
        }
 
+       if(artifactID == ArtifactID::SPELLBOOK)
+       {
+               hero->spells.insert(SpellID::SPELLBOOK_PRESET);
+               logGlobal->warn("Map '%s': Converted spellbook from artifact for hero %s.", mapName, hero->getNameTranslated().c_str());
+               return false;
+       }
+
        if(art->isBig() && slot >= ArtifactPosition::BACKPACK_START)
        {
                logGlobal->warn("Map '%s': A big artifact (war machine) in hero's backpack, ignoring...", mapName);
